### PR TITLE
Free tansfered memory when transferMemoryToContext fails

### PR DESCRIPTION
**Pull request by miniway**
I found a case the `general` memory pool was not recovered to initial value even though no query was running. Heap dump of the memory pool had a query entry which had raised `local memory limit` at  
```
    "stack": [
      "com.facebook.presto.ExceededMemoryLimitException.exceededLocalLimit(ExceededMemoryLimitException.java:34)",
      "com.facebook.presto.memory.QueryContext.reserveMemory(QueryContext.java:68)",
      "com.facebook.presto.operator.TaskContext.reserveMemory(TaskContext.java:153)",
      "com.facebook.presto.operator.SharedLookupSource.<init>(SharedLookupSource.java:40)",
      "com.facebook.presto.operator.ParallelHashBuilder$ParallelHashBuilderOperator.finish(ParallelHashBuilder.java:394)",
      "com.facebook.presto.operator.Driver.processInternal(Driver.java:358)",
      "com.facebook.presto.operator.Driver.processFor(Driver.java:303)",
      "com.facebook.presto.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:587)",
      "com.facebook.presto.execution.TaskExecutor$PrioritizedSplitRunner.process(TaskExecutor.java:505)",
      "com.facebook.presto.execution.TaskExecutor$Runner.run(TaskExecutor.java:640)",
      "java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)",
      "java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)",
      "java.lang.Thread.run(Thread.java:745)"
    ]
```
I feel like we might need to free transferred `TaskContext` memory when subsequent memory reservation fails. 

Also I found several zero-value entries in the MemoryPool.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java [hunk 1]
```diff
@@ -283,6 +284,12 @@ public void freeMemory(long bytes)
         memoryReservation.getAndAdd(-bytes);
     }
 
+    public void freeTaskContextMemory(long bytes)
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java]: Lets remove this method, and just have `SharedLookupSource` call `TaskContext.freeMemory`

File: presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java [hunk 2]
```diff
@@ -296,11 +303,24 @@ public void moreMemoryAvailable()
         memoryFuture.get().set(null);
     }
 
-    public long transferMemoryToTaskContext()
+    public void transferMemoryToTaskContext(long newMemoryBytes)
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java]: lets name the parameter `taskBytes` instead of `newMemoryBytes`

**Comment by cberner** [comment]
Thanks for finding that! Had a couple minor comments, but otherwise it looks good

**Comment by miniway** [comment]
Thanks for the comment. I've revised and rebased on the latest master.

**Comment by cberner** [comment]
Merged. Thanks!

**Pull request closed without merging**